### PR TITLE
Lombok

### PR DESCRIPTION
**Pull request by ranjanih**
Please read through the [writing guide](https://reflectoring.io/contribute/writing-guide/) before submitting a pull request and go through this checklist (at least for your first article).

**Feel free to raise the pull request even if the checklist hasn't been worked through, yet**, to take advantage of the online preview that is generated for each pull request (see "deploy/netlify" under "Show all checks" in your pull request). But please only flag it to be reviewed by your editor after going through the checklist.

## Review
- [ ] you have reviewed the text a day after finishing it (you'd be surprised how many errors you can avoid by just getting some distance to what you've written)

## Content
- [ ] the article explains *why* and not only *how*
- [ ] (half-)sentences that express a meaningful idea that is understandable by itself own are **bold** to allow quick scanning of the text
- [ ] the text has been checked with [Grammarly](https://grammarly.com) (the free tier is enough)

## Text
- [ ] the text is conversational ("we" instead of "you", easy language)
- [ ] there are no "walls of text" (short sentences, short paragraphs with a single idea)
- [ ] the text is inclusive ("the developers" and "they" instead of "the developer" and "he/she")
- [ ] the text uses active voice instead of passive voice
- [ ] names are spelled consistently throughout the article (correct uppercase / lowercase)
- [ ] file names and variables are highlighted as `code`

## Structure
- [ ] the article has a short intro before the first headline
- [ ] by reading the table of contents (TOC) alone, I get a sense of the "story" of the article (check the TOC in the preview after raising the PR or by starting the blog locally on your machine)
- [ ] if using sub-headlines, there's at least two sub-headlines below a main-headline
- [ ] headlines are consistent throughout the article (same level of abstraction, same style - if one headline starts with a verb in imperative form, the other headlines on the same level should probably, too)

## Code Examples
- [ ] code examples are introduced with a ":" and explained below the code example
- [ ] code examples are formatted to avoid horizontal scrolling (manual line breaks and 2-spaces indentation)

### CODE REVIEW

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+        private String dateOfBirth;
+        private String email;
+        private String phoneNo;
+        private String drivingLicenseNo;
+        private String licenseIssueState;
+    }
+ ````
+The above class has the following **flaws**:
+1. It is a **poorly designed model class**. We could consider restructuring the fields to create separate classes to store Address and LicenseDetails.
+2. **@Builder hides the complexity of autowiring instances** especially if custom objects
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md]: I'm not sure I understand this point. What is the "complexity of autowiring instances"? Can you elaborate? How does it tie into this section's title about "hiding validations"?

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+        private String drivingLicenseNo;
+        private String licenseIssueState;
+    }
+ ````
+The above class has the following **flaws**:
+1. It is a **poorly designed model class**. We could consider restructuring the fields to create separate classes to store Address and LicenseDetails.
+2. **@Builder hides the complexity of autowiring instances** especially if custom objects 
+having multiple dependencies are added to the class. This affects code readability and destroys the concept of the Single Responsibility Principle.
+3. If you have an application that uses **checkstyle/Sonar**, using these **annotations can hide errors and warnings** which is not desirable.
+This is shown in the sample below:
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md:403]: The images show the validation error, but in the sentence above you say that validation errors are not shown. This is confusing.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 3]
```diff
@@ -0,0 +1,474 @@
...
+   {{% image alt="delombok example" src="images/posts/lombok/checkstyle.png" %}}
+   {{% image alt="delombok example" src="images/posts/lombok/checkstyle_err.png" %}}
+4. With **`@AllArgsConstructor`**  referring to multiple same-type parameters, it is easy to accidentally define parameters out of order.
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md]: This doesn't have to do with "hiding validations" (which is the title of this section), does it? It feels like this point should move into the next section, because it's about a dependency to generated code.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+          processFile();
+       }
+   
+       private void processFile() throws IOException {
+           File file = new File("sample.txt");
+           throw new IOException();
+       }
+   }
+````
+As we can see, this **avoids the hassle of catching or throwing checked exceptions**. In other words, it treats a checked exception like an unchecked one.
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md]: The code without Lombok above just swallows (and logs) the exception! Looking at the docs, Sneakythrows rethrows the exception, however.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 5]
```diff
@@ -0,0 +1,474 @@
...
+This can be useful, especially when writing lambda functions making the code concise and clean. However, since the annotation swallows the checked exception we cannot catch them explicitly.
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md:466]: Ah, so `@SneakyThrows` also swallows the exception? I'm confused between this sentence and the one before. What does it do? 

If it just swallows the exception, the annotation is useless, in my opinion, so the section title should be "`@SneakyThrows` is Evil", not "`@SneakyThrows` Can be Evil"...

EDIT: I looked it up in the docs and SneakyThrows rethrows the exception, so the code example is wrong.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 6]
```diff
@@ -0,0 +1,474 @@
...
+It is easy to lose track of dependent objects with Lombok.
+4. Since @Builder gives a lot of flexibility in object creation it **can cause objects to be in an invalid state**. 
+Therefore, make sure all the required attributes are assigned values during object creation.
+5. **DO NOT write code that could have a huge dependency on the background code Lombok generates**.
+6. When using test coverage tools like Jacoco, Lombok can cause problems since **Jacoco cannot distinguish between Lombok generated code and normal source code**. 
+You might want to consider **excluding Lombok generated code for Jacoco test coverage**. More information on this is available [here](https://github.com/jacoco/jacoco/pull/495)
+7. **Use @SneakyThrows for checked exceptions that you don't intend to selectively catch**. Otherwise, wrap them in runtime exceptions that you throw instead.
+8. **Overusing @SneakyThrows** in an application could make it **difficult to trace and debug errors**.
+
+## Conclusion
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md]: I removed the conclusion, because the previous section is a great conclusion already.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+**Ensure it is used cautiously and not as an alternative to bypass checked exceptions.**
+
+## Use Lombok with caution
+
+The power of Lombok cannot be underestimated or ignored. However, I would like to summarise the key points
+that will help you use Lombok in a better way.
+1. **Avoid using Lombok with entities**. It will be much easier generating the code yourself than debugging issues later.
+2. When designing POJO's **use only the Lombok annotations you require**. (Use shorthand annotations wisely).
+I would recommend using the Delombok feature to understand the code generated better.
+3. **Do not add too many dependencies in the POJO's**. Keep the classes relevant to the responsibility they are designed for.
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md]: I'm not sure what this has to do with Lombok. This is good advice in general, but how does it tie in with Lombok?

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+           File file = new File("sample.txt");
+           throw new IOException();
+       }
+   }
+````
+As we can see, this **avoids the hassle of catching or throwing checked exceptions**. In other words, it treats a checked exception like an unchecked one.
+This can be useful, especially when writing lambda functions making the code concise and clean. However, since the annotation swallows the checked exception we cannot catch them explicitly. 
+Instead, we must handle the exceptions via a global exception handler.
+Therefore, **use it only when you don't intend to process the code selectively depending on the kind of Exception it throws**.
+**Ensure it is used cautiously and not as an alternative to bypass checked exceptions.**
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md:470]: The main feature of `SneakyThrows` is to bypass checked exceptions, so I don't understand this sentence.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 9]
```diff
@@ -0,0 +1,474 @@
...
+## Use Lombok with caution
+
+The power of Lombok cannot be underestimated or ignored. However, I would like to summarise the key points
+that will help you use Lombok in a better way.
+1. **Avoid using Lombok with entities**. It will be much easier generating the code yourself than debugging issues later.
+2. When designing POJO's **use only the Lombok annotations you require**. (Use shorthand annotations wisely).
+I would recommend using the Delombok feature to understand the code generated better.
+3. **Do not add too many dependencies in the POJO's**. Keep the classes relevant to the responsibility they are designed for. 
+It is easy to lose track of dependent objects with Lombok.
+4. Since @Builder gives a lot of flexibility in object creation it **can cause objects to be in an invalid state**.
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md]: Good point! I don't recall that you have mentioned that in the builder section above. Can you do that?

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 10]
```diff
@@ -0,0 +1,474 @@
...
+that will help you use Lombok in a better way.
+1. **Avoid using Lombok with entities**. It will be much easier generating the code yourself than debugging issues later.
+2. When designing POJO's **use only the Lombok annotations you require**. (Use shorthand annotations wisely).
+I would recommend using the Delombok feature to understand the code generated better.
+3. **Do not add too many dependencies in the POJO's**. Keep the classes relevant to the responsibility they are designed for. 
+It is easy to lose track of dependent objects with Lombok.
+4. Since @Builder gives a lot of flexibility in object creation it **can cause objects to be in an invalid state**. 
+Therefore, make sure all the required attributes are assigned values during object creation.
+5. **DO NOT write code that could have a huge dependency on the background code Lombok generates**.
+6. When using test coverage tools like Jacoco, Lombok can cause problems since **Jacoco cannot distinguish between Lombok generated code and normal source code**.
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md:484]: Good point! You haven't mentioned JaCoCo above! This might be worth to mention in the section about validations (maybe rephrase the title of that section to something like "Using Lombok with Static Code Analyzers"

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+I would recommend using the Delombok feature to understand the code generated better.
+3. **Do not add too many dependencies in the POJO's**. Keep the classes relevant to the responsibility they are designed for. 
+It is easy to lose track of dependent objects with Lombok.
+4. Since @Builder gives a lot of flexibility in object creation it **can cause objects to be in an invalid state**. 
+Therefore, make sure all the required attributes are assigned values during object creation.
+5. **DO NOT write code that could have a huge dependency on the background code Lombok generates**.
+6. When using test coverage tools like Jacoco, Lombok can cause problems since **Jacoco cannot distinguish between Lombok generated code and normal source code**. 
+You might want to consider **excluding Lombok generated code for Jacoco test coverage**. More information on this is available [here](https://github.com/jacoco/jacoco/pull/495)
+7. **Use @SneakyThrows for checked exceptions that you don't intend to selectively catch**. Otherwise, wrap them in runtime exceptions that you throw instead.
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md]: Good point! This point didn't come across very well in the section about SneakyThrows above. Now I understand what you mean! Please make this point clear in the SneakyThrows section. Add another code example of an exception handler, maybe, to help clarify this.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 12]
```diff
@@ -0,0 +1,474 @@
...
+8. **Overusing @SneakyThrows** in an application could make it **difficult to trace and debug errors**.
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md:487]: Good point. How so, though? I guess because the code that throws the exception doesn't exist in our codebase. This point also didn't come across very well in the SneakyThrows section above.

**Review by thombergs — CHANGES_REQUESTED** [review]
Good language and structure! I left some comments (especially on the section with the caveats) where some things weren't quite clear to me. If you can rephrase/restructure that section, we'll have a good article on our hands!

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+        // All getters and setters are explicitly defined here    
+    
+        public String toString() {
+            return "Book(isbn=" + this.getIsbn() + ", publication=" + this.getPublication() + ", title=" + this.getTitle() + ", authors=" + this.getAuthors() + ", genre=" + this.getGenre() + ")";
+        }
+    }
+
+ ````
+
+Lombok simplifies the above **plain Java class** to this :
```
[the remark this answers is not in the document]
Comment (ranjanih) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md]: Thank you for this!

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md (hunk 10 above)
Comment (ranjanih) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md:484]: @thombergs  I see that the website already has an aricle about Jacoco that mentions the steps to exclude the code generated by Lombok. Do I just add a link to this in the main section or should I add some details on why this is reqd and then add a link to this section?
https://reflectoring.io/jacoco/#excluding-code-generated-by-lombok

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md (hunk 11 above)
Comment (ranjanih) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md]: @thombergs I think I have a fair amount of idea on what went wrong here. I will try to rephrase and simplify the SneakyThrows section. Thanks for pointing out.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md (hunk 9 above)
Comment (ranjanih) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md]: Noted.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+
+>  "Project Lombok is a java library that automatically plugs into your editor and build tools, spicing up your Java."
+
+This library comprises a set of user-friendly annotations that generate the code at compile-time, helping the developers **save time, space and improving code readability**.
+
+## IDE support 
+
+All popular IDEs support Lombok via a plugin. We need to ensure annotation processing is enabled as below.
+{{% image alt="settings" src="images/posts/lombok/settings.PNG" %}}
+They help the IDE generate the source at compile-time and helps us with code debugging.
```
[the remark this answers is not in the document]
Comment (ranjanih) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md]: Sure. will add them.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+This article is accompanied by working code examples that demonstrates commonly used Lombok features on [Github.](https://github.com/thombergs/code-examples)
+
+
+## What is Lombok
+
+According to official docs,
+
+>  "Project Lombok is a java library that automatically plugs into your editor and build tools, spicing up your Java."
+
+This library comprises a set of user-friendly annotations that generate the code at compile-time, helping the developers **save time, space and improving code readability**.
```
[the remark this answers is not in the document]
Comment (ranjanih) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md]: Got your point. Thanks

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md (hunk 1 above)
Comment (ranjanih) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md]: @thombergs  I mainly want to highlight 4 points
1. Sonar does not catch the violations when annotations are added.
2. A class having a huge constructor means something is wrong. Maybe the class has too much responsibility.
3. Using the right combination of @Builder and @RequiredArgsConstructor is essential so that the object created is not in an invalid state.
4. If a constructor has too many same type params. eg. too many String params, it is easy to define the object where the params can get out of order which might be difficult to trace.

Need some help structuring here. I am not sure of each of these points should be a separate heading or should i combine some of them together.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+        private Gender gender;
+        private String dateOfBirth;
+        private String email;
+        private String phoneNo;
+        private String drivingLicenseNo;
+        private String licenseIssueState;
+    }
+ ````
+The above class has the following **flaws**:
+1. It is a **poorly designed model class**. We could consider restructuring the fields to create separate classes to store Address and LicenseDetails.
```
[the remark this answers is not in the document]
Comment (ranjanih) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md]: Yes. will remove it and keep the content relevant to the heading.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md (hunk 1 above)
Comment (ranjanih) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md]: Also will try to restructure the validation better by showing before Lombok result and after Lombok result.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md (hunk 10 above)
Comment (thombergs) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md:484]: Just mention it and add a link

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md (hunk 1 above)
Comment (thombergs) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md]: I suggest to structure it like this:

* section "Lombok Hides Coding Violations"
  * explains checkstyle etc. and how Lombok hides their violations
* section "`@AllArgsConstructor` May Introduce Errors When Refactoring"
  * explains that the order of constructor parameters changes when you change the order of fields
* section "`@Builder` Allows Creation of Invalid Objects
  * explains how `@Builder` lets us create objects in invalid states
 
each section with its own small code example and explanation.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 17]
```diff
@@ -0,0 +1,552 @@
...
+        CARPENTER
+    }
+```
+For this class, we could construct an object as 
+```java
+    Job job = Job.builder().id("5678").build();
+```
+Although, the code compiles, the object `job` here is in an invalid state because we do not know which `JobType` it belongs to.
+Therefore, along with using the @Builder annotation, it is also important to enforce required attributes to have a value.
+To do this we could consider using **@NonNull** annotation.
```
Comment (thombergs) [inline content/blog/2022/2022-03-31-be-informed-with-lombok.md]: Can you add a sentence explaining whether it's possible to add custom validations in some way? And maybe add a link to the validation annotations that Lombok supports aside from `NonNull`?

**Review by thombergs — CHANGES_REQUESTED** [review]
Two more small comments, then we're ready to publish!

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md (hunk 17 above)
Comment (ranjanih) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md:461]: There are no other validations in lombok. I have added a reference to the Bean validation API here.

File: content/blog/2022/2022-03-31-be-informed-with-lombok.md [hunk 18]
```diff
@@ -0,0 +1,552 @@
...
+This library provides a set of user-friendly annotations that generate the code at compile-time, helping the developers **save time, space and improving code readability**.
+
+## IDE Support 
+
+All popular IDEs support Lombok. For example, *Intellij* version 2020.3 and above is compatible with Lombok without a plugin. For earlier versions, plugins can be installed from [here](https://plugins.jetbrains.com/plugin/6317-lombok/).
+Once installed, we need to ensure annotation processing is enabled as in the example configuration below.
+{{% image alt="settings" src="images/posts/lombok/settings.PNG" %}}
+Annotation processing makes it possible for the IDE to evaluate the Lombok annotations and generate the source code from them at compile-time.
+
+For *Eclipse*, install the Lombok plugin downloaded from this [location](https://search.maven.org/search?q=g:org.projectlombok%20AND%20a:lombok&core=gav).
```
[the remark this answers is not in the document]
Comment (ranjanih) [reply content/blog/2022/2022-03-31-be-informed-with-lombok.md]: Done.
